src/mono/wasm/runtime/hybrid-globalization.ts

```diff
@@ -135,24 +135,115 @@ export function mono_wasm_ends_with(exceptionMessage: Int32Ptr, culture: MonoStr
     }
 }
 
-function get_clean_string(strPtr: number, strLen: number) {
-    const str = string_decoder.decode(<any>strPtr, <any>(strPtr + 2 * strLen));
+function decode_to_clean_string(strPtr: number, strLen: number)
+{
+    const str = string_decoder.decode(<any>strPtr, <any>(strPtr + 2*strLen));
+    return clean_string(str);
+}
+
+function clean_string(str: string)
+{
     const nStr = str.normalize();
     return nStr.replace(/[\u200B-\u200D\uFEFF\0]/g, "");
 }
 
+export function mono_wasm_index_of(exceptionMessage: Int32Ptr, culture: MonoStringRef, needlePtr: number, needleLength: number, srcPtr: number, srcLength: number, options: number, fromBeginning: number): number{
+    const cultureRoot = mono_wasm_new_external_root<MonoString>(culture);
+    try {
+        const needle = string_decoder.decode(<any>needlePtr, <any>(needlePtr + 2*needleLength));
+        // no need to look for an empty string
+        if (clean_string(needle).length == 0)
+            return fromBeginning ? 0 : srcLength;
+
+        const source = string_decoder.decode(<any>srcPtr, <any>(srcPtr + 2*srcLength));
+        // no need to look in an empty string
+        if (clean_string(source).length == 0)
+            return fromBeginning ? 0 : srcLength;
+        const cultureName = conv_string_root(cultureRoot);
+        const locale = cultureName ? cultureName : undefined;
+        const casePicker = (options & 0x1f);
+
+        const segmenter = new Intl.Segmenter(locale, { granularity: "grapheme" });
+        const needleSegments = Array.from(segmenter.segment(needle)).map(s => s.segment);
+        let i = 0;
+        let stop = false;
+        let result = -1;
+        let segmentWidth = 0;
+        let index = 0;
+        let nextIndex = 0;
+        while (!stop)
+        {
+            const iteratorSrc = segmenter.segment(source.slice(i, source.length))[Symbol.iterator]();
+            let srcNext = iteratorSrc.next();
+
+            if (srcNext.done)
```

Comment:
When could this happen ?

Reply:
When we are over the `index = sourceSegements.Lenght - targetSegments.Length` in the source string.
E.g. being on the `3` char of source: `123` looking for target: `34`. We have a match, needle still has the next value but source is out of values.
We don't have the access to source lenght (like we do to `needleSegments.length`) so we cannot do early stop condition when we are at the mentioned `index`. And `source. length` is not the same as `sourceSegements.Lenght`.